helper/resource/teststep_providers_test.go

```diff
@@ -342,6 +344,66 @@ func TestTest_TestStep_ExternalProviders_To_ProviderFactories_StateUpgraders(t *
 	})
 }
 
+func TestTest_TestStep_Taint(t *testing.T) {
+	t.Parallel()
+
+	var idOne, idTwo string
+
+	Test(t, TestCase{
+		ExternalProviders: map[string]ExternalProvider{
+			"random": {
+				Source:            "registry.terraform.io/hashicorp/random",
+				VersionConstraint: "3.3.2",
+			},
+		},
+		Steps: []TestStep{
+			{
+				Config: `resource "random_string" "test" {
+					length = 10
+				}`,
+				Check: ComposeAggregateTestCheckFunc(
+					TestCheckResourceAttr("random_string.test", "length", "10"),
+					ExtractResourceAttr("random_string.test", "id", &idOne),
+				),
+			},
+			{
+				Taint: []string{"random_string.test"},
+				Config: `resource "random_string" "test" {
+					length = 10
+				}`,
+				Check: ComposeAggregateTestCheckFunc(
+					TestCheckResourceAttr("random_string.test", "length", "10"),
+					ExtractResourceAttr("random_string.test", "id", &idTwo),
+				),
+			},
+		},
+	})
+
+	if idOne == idTwo {
+		t.Errorf("taint is not causing destroy-create cycle, idOne == idTwo: %s == %s", idOne, idTwo)
+	}
+}
+
+func ExtractResourceAttr(resourceName string, attributeName string, attributeValue *string) TestCheckFunc {
```

Comment:
Perhaps we could move this to a non-test file so that it's available for provider developers?

Reply:
Let's create a followup issue to gauge interest 👍 

Although we may want be careful to unexport this function here since it could easily wind up somewhere where it could be added to the public API.

```suggestion
func extractResourceAttr(resourceName string, attributeName string, attributeValue *string) TestCheckFunc {
```